src/extension_instance.ml

```diff
@@ -0,0 +1,143 @@
+       We can use versions instead of capabilities to suggest the user to update their
+       ocaml-lsp. *)
+  then
+    show_message `Warn
+      "The installed version of ocamllsp is out of date. Some features may be \
+       unavailable or degraded in functionality: switching between \
+       implementation and interface files, functionality in mli sources. \
+       Consider updating ocamllsp.";
+
+  Ok (client, ocaml_lsp)
+
+let restart_language_server t =
+  let open Promise.Result.Syntax in
+  LanguageClient.stop t.client;
+  let+ language_client, ocaml_lsp = start_language_server t.toolchain in
+  t.client <- language_client;
+  t.ocaml_lsp <- ocaml_lsp
+
+module Sandbox_info : sig
+  val make : Toolchain.t -> StatusBarItem.t
+
+  val update : StatusBarItem.t -> new_toolchain:Toolchain.t -> unit
+end = struct
+  let make_status_bar_item_text package_manager =
+    Printf.sprintf "%s %s" LabelIcons.package
+    @@ Toolchain.Package_manager.to_pretty_string package_manager
+
+  let make toolchain =
+    let status_bar_item =
+      Window.createStatusBarItem ~alignment:StatusBarAlignment.Left ()
+    in
+    let status_bar_item_text =
+      make_status_bar_item_text @@ Toolchain.package_manager toolchain
+    in
+    StatusBarItem.set_text status_bar_item status_bar_item_text;
+    StatusBarItem.set_command status_bar_item
+      (`String Extension_consts.Commands.select_sandbox);
+    StatusBarItem.show status_bar_item;
+    status_bar_item
+
+  let update sandbox_info ~new_toolchain =
+    let status_bar_item_text =
+      make_status_bar_item_text @@ Toolchain.package_manager new_toolchain
+    in
+    StatusBarItem.set_text sandbox_info status_bar_item_text
+end
+
+let make toolchain =
+  let sandbox_info = Sandbox_info.make toolchain in
+  let open Promise.Result.Syntax in
+  let+ client, ocaml_lsp = start_language_server toolchain in
+  { toolchain; client; ocaml_lsp; sandbox_info }
+
+let update_on_new_toolchain t new_toolchain =
+  Sandbox_info.update t.sandbox_info ~new_toolchain;
+  LanguageClient.stop t.client;
+  let open Promise.Result.Syntax in
+  let+ client, ocaml_lsp = start_language_server new_toolchain in
+  t.toolchain <- new_toolchain;
+  t.client <- client;
+  t.ocaml_lsp <- ocaml_lsp;
+  ()
+
+let open_terminal toolchain =
+  match Terminal_sandbox.create toolchain with
+  | Some terminal -> Terminal_sandbox.show terminal
+  | None ->
+    show_message `Error
+      "Could not open a terminal in the current sandbox. The toolchain may not \
+       have loaded yet."
+
+let disposable t =
+  Disposable.make ~dispose:(fun () ->
+      (* [stop] is not defined in the toplevel module because
+         we don't want anyone to call it, only vscode when the extension is no longer needed;
+         a user calling [stop] would leave the extension in a corrupt state *)
+      let stop instance =
+        StatusBarItem.dispose instance.sandbox_info;
+        LanguageClient.stop instance.client
+      in
+      stop t)
```

Comment:
I think it's unnecessary to define a local function inside of a lambda that's only used once.

```suggestion
  Disposable.make ~dispose:(fun () ->
      StatusBarItem.dispose instance.sandbox_info;
      LanguageClient.stop instance.client)
```

Reply:
For readability :-) Feel free to commit suggestion if you want